In [1]:
%pip install surprise

     -------------------------------------- 772.0/772.0 kB 9.8 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Running setup.py install for scikit-surprise: started
  Running setup.py install for scikit-surprise: finished with status 'done'
Note: you may need to restart the kernel to use updated packages.


  DEPRECATION: scikit-surprise is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/8559

[notice] A new release of pip available: 22.3 -> 22.3.1
[notice] To update, run: C:\Users\Hcoli\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [3]:
from surprise.model_selection import GridSearchCV, cross_validate
from surprise import Dataset, NMF
import numpy as np

In [7]:
data = Dataset.load_builtin(name = "jester", )
param_grid = {"n_factors": range(4, 30, 10), "n_epochs": range(30, 50, 10), "biased" : [True, False], "reg_pu": np.arange(0.05, 0.15, 0.05), "reg_qi" : np.arange(0.05, 0.15, 0.05)}
gs = GridSearchCV(NMF, param_grid, measures=['rmse'], cv=5)
gs.fit(data)

print(gs.best_score['rmse'])

print(gs.best_params['rmse'])

4.300627109814924
{'n_factors': 4, 'n_epochs': 30, 'biased': True, 'reg_pu': 0.1, 'reg_qi': 0.05}


PROJECT SUMMARY

In this project, we used the Surprise library to create a joke recommendation system. We used the jester dataset that's available in the Surprise library which consists of 6.5 million anonymous ratings of jokes from over 70,000 users. We first performed 5 fold cross validation to tune our hyperparameters. We chose 5 hyperparameters to tune: the number of latent dimensions in our model, the number of epochs used for SGD, whether or not our model should use biases, and regularization parameters for pu, which represents the user's preferences, and qi, which represents the characteristics of the joke. 

The number of latent dimensions is the dimension of the space that the vectors pu and qi live in. We found that the optimal number of dimensions found by the model was only 4, our lowest value, which is really interesting. We're kind of curious to know what attributes the model ended up choosing.

The number of epochs used is the number of passes over the training set that the algorithm makes in updating the weight vector. If the number of epochs is too large, then it's possible that the model might overfit on the data. Additionally, it's possible that the weight vector converges to a local minimum well before the total number of epochs is reached and so any additional iterations might end up being unncecessary and inefficient. Cross validation showed that the optimal number of epochs was 30. This was the lowest value number of epochs that we had the model try out so it's possible that an even smaller number of epochs can be used.

We thought it was interesting that one of the parameters was for whether or not the model should incorporate biasing. We weren't sure why you wouldn't want to use biasing so we kept it as a parameter to see what would happen. 

Finally, for the regularization parameters, we found that the best values were 0.1 for the rating vector and 0.05 for the item vector. Initially, we had wanted to sweep the regularization parameters from 0 to 1 in increments of 0.1 but found that this made cross validation take too long. So we instead had the values range from 0.05 to 0.15 since we believed that our model would likely pick optimal values close to the recommended default regularization value of 0.06. 

Our RMSE after tuning our parameters was roughly 4.3. Since the jester dataset has ratings going from -10 to 10, which gives a normalized RMSE of 0.215 which is a fairly reasonable result.